README

OVERVIEW
--------
GNU Unifont is an official GNU package.  It is a dual-width
(8x16/16x16) bitmap font, designed to provide coverage for
all of Unicode Plane 0, the Basic Multilingual Plane (BMP).

This version has a glyph for each visible code point in the
Unicode 6.3 Basic Multilingual Plane (Plane 0) and some glyphs
in the Supplemental Multilingual Plane (Plane 1).  This version
also includes many glyphs in Michael Everson's ConScript
Unicode Registry (CSUR).

Unifont only provides a single glyph for each character, making it
impossible to handle any language properly that needs context-dependent
character shaping. It is supplied in the form of a hex file, with
a converter to convert it to BDF. See http://czyborra.com/unifont/
or http://unifoundry.com/unifont.html for more information.  The
BDF font is converted to PCF, and the hex file is converted to a
TrueType font.

This is the unifoundry.com collection of utilities for GNU Unifont,
assembled by Paul Hardy with the encouragement of the font's creator,
Roman Czyborra.  This archive contains the following directories
and files:

     ChangeLog  Log of changes made to each GNU release
     COPYING    Full text of GPL version 2
     doc        Documentation in Texinfo format
     font       The font source file with scripts for building
     hangul     Standalone font sources to build hangul-syllables.hex
     INSTALL    Instructions for font and software installation
     Makefile   The "make" file
     man        Unix man pages
     NEWS       Summary of what's new with each GNU release
     README     This file
     src        Source programs, in Perl and C

The "font/precompiled" directory contains prebuilt font-related files:

     coverage.txt                     Percentage coverage of Plane 0 scripts

     unifont-<version>.hex            Hex string source of glyphs to build
                                      Unifont
     unifont-<version>.bdf.gz         BDF version of Unifont
     unifont-<version>.pcf.gz         PCF version of Unifont
     unifont-<version>.ttf            TrueType version of Unifont

     unifont_sample-<version>.hex     Hex string source of all Plane 0 glyphs
                                      (except those for U+FFFE and U+FFFF),
                                      including nonprinting and PUA glyphs,
                                      with combining circles
     unifont_sample-<version>.bdf.gz  BDF font version of the above .hex file
     unifont_sample-<version>.ttf     SBIT font version of the above .hex file

     unifont_csur-<version>.*         Fonts containing Plane 0 Unifont glyphs
                                      plus glyphs for Michael Everson's
                                      ConScript Unicode Registry (CSUR) for
                                      the Plane 0 Private Use Area

     unifont_upper-<version>.*        Fonts containing glyphs from Unicode
                                      Plane 1 through Plane 14, inclusive

     unifont_upper_csur-<version>.*   Fonts containing glyphs from Unifont
                                      Upper plus glyphs from Michael Everson's
                                      ConScript Unicode Registry (CSUR) that
                                      are in the Private Use Area in Plane 15

     unifont-<version>.bmp            The entire Plane 0 Unifont font with
                                      combining circles, built from the files
                                      font/plane00/*.hex, showing combining
                                      circles

The directory that was originally named "font/hexsrc" has been renamed
to "font/plane00" now that Unifont supports glyphs beyond Plane 0.  Higher
plane glyphs appear in "font/plane01" through "font/plane0F".  Currently
there is no "font/plane10" directory (the highest Unicode plane is Plane 17,
or 0x10).

This release incorporates all glyph errata issued by The Unicode Consortium
from Unicode 1.0 errata to the latest.


BUILDING
--------
See the "INSTALL" file in this directory for building instructions.


src/ AUTHORS
------------
Roman Czyborra wrote all the Perl files in the src directory except
"hex2sfd", "hexkinya", "unifontchojung", "unifontksx", "unihex2png",
and "unipng2hex".

In the case of "johab2ucs2", Jungshik Shin wrote the orignial version;
he then gave it to Roman.  Paul Hardy made further changes to "johab2ucs2".

Roman originally named the "src/hexbraille" script as simply "braille".
Paul Hardy thought there was too great a chance of a name conflict with
other utilities, and so renamed it.

Luis Alejandro Gonzalez Miranda wrote the original "hex2sfd" Perl
script, as well as a "howto-build.sh" shell script that Paul Hardy
converted into "./font/ttfsrc/Makefile".

Paul Hardy wrote "unifontchojung" and "unifontksx" for extracting subsets
of Hangul glyphs, as an aid in creating a new Hangul Syllables block.

Andrew Miller wrote "hexkinya".  He also wrote "unihex2png" and "unipng2hex"
based upon Paul Hardy's "unihex2bmp" and "unibmp2hex" programs.

Paul Hardy wrote all the C programs.


Unifont AUTHORS
---------------
Roman Czyborra created the original GNU Unifont, including the
.hex format.  For greater detail, see the HISTORY section below.

David Starner aggregated many glyphs contributed by others and
incorporated these into pre-2004 Unifont releases.

Qianqian Fang began his Wen Quan Yi font in 2004, by which
time work on Unifont had stopped.  Most of the almost 30,000
CJK ideographs in Unifont versions 5.1 and later were taken
from Wen Quan Yi with permission of Qianqian Fang.  The glyphs
in "./font/hexsrc/wqy-cjk.hex" are for the most part Qianqian
Fang's Unibit and Wen Quan Yi glyphs.

Paul Hardy drew most of the newly-drawn glyphs added to the BMP
from the Unifont 5.1 release to the present release.  This includes
the 11,172 glyphs in the Hangul Syllables block, plus approximately
10,000 additional glyphs scattered throughout the BMP.

Andrew Miller drew the glyphs added to Unicode 6.3.0.

For higher planes and the Private Use Area glyphs, see the ChangeLog
file.


LICENSE
-------
The source code for everything except the compiled fonts in this current
release is licensed as follows:

     License for this current distribution of program source
     files (i.e., everything except the fonts) is released under
     the terms of the GNU General Public License version 2,
     or (at your option) a later version.

GPL version 2 is contained in the "COPYING" file in the main source
directory for this package.  If your received this source without
a copy of GPL version 2, you can download a copy from GNU's website
at http://www.gnu.org/licenses/gpl-2.0.html.

The license for the compiled fonts is covered by the above GPL terms
with the GNU font embedding exception, as follows:

     As a special exception, if you create a document which uses this font,
     and embed this font or unaltered portions of this font into the document,
     this font does not by itself cause the resulting document to be covered
     by the GNU General Public License. This exception does not however
     invalidate any other reasons why the document might be covered by the
     GNU General Public License. If you modify this font, you may extend
     this exception to your version of the font, but you are not obligated
     to do so. If you do not wish to do so, delete this exception statement
     from your version. 

See "http://www.gnu.org/licenses/gpl-faq.html#FontException" for more details.


CHANGES IN VERSION 6.3
----------------------
Version 6.3 reflects all glyph changes and errata published in Unicode
6.3.0.  In preparation for releasing this version, Paul Hardy obtained
a hard copy of the errata published in Unicode Version 1.1, not yet
available on Unicode's website.  All previously published errata have
been incorporated.  This is a complete replacement for all previous
releases.

The following code points in previously published errata were examined
and found to be correct:

     Unicode 1.1: U+717F, U+773E, U+809C, U+8480, U+908E

Andrew Miller drew the 5 new additions to the Unicode 6.3.0 Basic
Multilingual Plane in the initial Unifont 6.3 release.

The latest Unifont 6.3 release includes these glyph changes by Paul Hardy:

   - Armenian -- several glyphs were redrawn based upon feedback from
     native speakers (U+0530..U+058F).

   - CJK Radicals Supplement -- several glyphs were redrawn to better match
     their representations in The Unicode Standard code charts:
     U+2E9F, U+2EA9, U+2EAC, U+2EAE, U+2EC0, U+2EDE, U+2EE7, and U+2EED.

   - Capricorn sign (U+2651) -- this was redrawn to an alternate form that
     better fit in an 8 by 16 pixel grid.

   - Dashes -- changed to distguish better between different dash types
     (a two horizontal pixel difference is the minimum to easily distinguish
     a difference between two glyphs):
     * Hyphen (U+002D) and Soft Hyphen (U+00AD) are now 4 pixels wide
     * En Dash (U+2012) is now 6 pixels wide
     * Em Dash (U+2013) is now 8 pixels wide

   - Control Pictures:
     * Centered text for C1 Controls U+0089 ("HTJ"), U+0095 ("MW"),
       and U+009E ("PM")
     * Copied glyphs from U+0000..U+001F to U+2400..u+241F and erased
       surrounding borders; earlier, some glyphs in U+0000..U+001F had
       their text re-centered so this carries that change forward

   - Arrows -- General Re-alignment
     * Aligned most single vertical arrow strokes with the 5th column,
       counting from the left, to align with the center of the "w" glyph
       (U+0077)
     * Aligned most single horizontal arrow strokes with the 7th row,
       counting from the bottom, to align with the horizontal stroke in
       the "e" glyph (U+0065); this follows the convention of Donald Knuth's
       fonts in TeX, as illustrated in The TeXbook
     * Modified the following ranges per the above two re-alignments:
       o U+2190..U+21FF Arrows
       o U+27F0..U+27FF Supplemental Arrows -- A
       o U+2900..U+297F Supplemental Arrows -- B
       o U+2B00..U+2BFF Miscellaneous Symbols and Arrows

   - Modified the following additional Miscellaneous Technical glyphs
     * Scan lines for old 9-line character terminals:
       o U+23BA Line 1, horizontal line across row  1 (counting from the top)
       o U+23BB Line 3, horizontal line across row  5 (counting from the top)

NEWS
2017-08-27: Release 10.0.06

   Added numerous improvements to Unicode Plane 0 and Plane 1 scripts,
   almost all of which were contributed by David Corbett; see the
   ChangeLog file for details.

   Made changes to make files contributed by Mike Gilbert to support
   parallel make.

   Fixed a corrupted ASCII hexadecimal string array in unifontpic.h.

2017-07-12: Release 10.0.05

   This release positions Unicode double diacritic marks correctly
   in the TrueType fonts.  The previous release did not multiply
   their offsets by the correct scale factor in FontForge.

   The unifontpic chart title is also centered along the top row
   legend, which looks better visually.

   The Phonetic Extensions block has gone through major revision,
   with minor adjustments in other Unicode scripts as described
   in the ChangeLog file.

2017-07-08: Release 10.0.04

   Added "-P" flag to the unifontpic utility to specify the Unicode
   plane.  This allows creating code charts for glyphs above Plane 0.
   The Unicode plane is now added to the chart title.  This works
   properly for the wide and the long chart versions, with the
   letters in the long version's chart title now single-spaced.

   Received double-width Chinese ideographs to replace the quadruple-
   width glyphs for Church Slavonic transliteration that were released
   in Unifont 10.0.01.  This allows a return to PNG charts of Unifont
   glyphs.  PNG files are smaller than the BMP version but the PNG
   Unifont utilities have not been modified to handle quadruple-width
   glyphs, whereas the BMP utilities have been.  The 18 quadruple-width
   Chinese glyphs are preserved in font/plane00/alt/quad-width.hex to
   allow future experimentation.

   The PNG charts of Unicode Plane 1 (Supplemental Multilingual Plane)
   glyphs now use the Unifont "sample" version of Unifont ".hex" files,
   so that combining circles appear.  This was already being done for
   Unicode Plane 0 and will look better in font sample pages.

   Added new x-offset field to all *-combining.txt files to provide
   proper rendering of Unicode double diacritic marks; they should
   span across the preceding character and the following character.
   Added support for this in hex2sfd for FontForge generation of
   TrueType fonts.  Also added support for this to the unigencircles
   utility, so the combining circle of a double diacritic is printed
   correctly in the left half of the glyph as in the Unicode
   Consortium's code charts, rather than being centered.  Several
   applications are not rendering these double diacritic marks
   correctly; releasing this version of Unifont will allow validation
   of the new Unifont TrueType build, and could serve as a test font
   for implementing these specialty Unicode characters in other
   applications.

   Redrew Musical Symbols Ornaments (U+01D194..U+01D1A5) so they
   join together when used as a consecutive string of characters.
   Adjusted the Byzantine Musical Symbols "ano", "meso", and "kato"
   glyph versions for high, middle, and low vertical placement,
   respectively.

   Made numerous adjustments to many glyphs as noted in the ChangeLog
   file, including re-coding some "subjoiner" glyphs as combining
   characters instead of as non-printing.  Those glyphs now render
   a virama (halant) sign if a script contains such a sign, or a
   "+" subscript sign otherwise.

2017-06-30: Release 10.0.03

   Modified font/hex2sfd to reposition combining characters.  If
   you experience problems with the TrueType font in this release,
   revert to 10.0.02.

2017-06-30: Release 10.0.02

   Touched up glyphs in the block U+03xx, which includes modern Greek.
   Fixed an Aiha glyph in the Private Use Area in the Unifont CSUR font.

2017-06-20: Release 10.0.01

   Updated for Unicode 10.0 release.

   New scripts added (introduced in Unicode 10.0):
     - U+0860..U+086F     Syriac Supplement
     - U+011A00..U+011A4F Zanabazar Square
     - U+011A50..U+011AAF Soyombo
     - U+011D00..U+011D5F Masaram Gondi

   Added quadruple-width glyphs U+9FD8..U+9FE9, updating some utility
   programs and documentation to allow this change.  The Perl scripts
   unihex2png, unipng2hex, and unifont-viewer do not support this new
   16-by-32 glyph size.

   Corrected several glyphs and made numerous changes in others,
   described in detail in the ChangeLog file.

2016-12-22: Release 9.0.06

   Refined about 50 Pikto CSUR glyphs.

   Corrected two Osage glyphs and one mathematical operator as noted
   in ChangeLog.

2016-12-10: Release 9.0.05

   Modified Cherokee glyphs to reflect changes in recent Cherokee font
   that the Unicode Consortium adopted for its code charts.  Removed
   serifs to make Cherokee glyphs easier to read.

   Modified Pikto CSUR glyphs after comparison with high-resolution scans
   of Basic Pikto book.

2016-10-29: Release 9.0.04

   Extended "{" and "}" by one pixel so they now come to a point in the center.
   Swapped U+2A05 and U+2A06 so they are now in the correct order.

2016-10-21: Release 9.0.03

   Added CSUR glyphs:
     - Aiha: added doubling combining glyph, omitted from the ConScript
       Unicode Registry (CSUR) code chart.

src/unihex2bmp.c
/*
   unihex2bmp - program to turn a GNU Unifont hex glyph page of 256 code
                points into a Microsoft Bitmap (.bmp) or Wireless Bitmap file

   Synopsis: unihex2bmp [-iin_file.hex] [-oout_file.bmp]
                [-f] [-phex_page_num] [-w]


   Author: Paul Hardy, unifoundry <at> unifoundry.com, December 2007
   
   Copyright (C) 2007, 2008, 2013 Paul Hardy

   LICENSE:

      This program is free software: you can redistribute it and/or modify
      it under the terms of the GNU General Public License as published by
      the Free Software Foundation, either version 2 of the License, or
      (at your option) any later version.

      This program is distributed in the hope that it will be useful,
      but WITHOUT ANY WARRANTY; without even the implied warranty of
      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
      GNU General Public License for more details.

      You should have received a copy of the GNU General Public License
      along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAXBUF 256


/*
   These are the GNU Unifont hex strings for '0'-'9' and 'A'-'F',
   for encoding as bit strings in row and column headers.

   Looking at the final bitmap as a grid of 32*32 bit tiles, the
   first row contains a hexadecimal character string of the first
   3 hex digits in a 4 digit Unicode character name; the top column
   contains a hex character string of the 4th (low-order) hex digit
   of the Unicode character.
*/
char *hex[18]= {
      "0030:00000000182442424242424224180000",  /* Hex digit 0 */
      "0031:000000000818280808080808083E0000",  /* Hex digit 1 */
      "0032:000000003C4242020C102040407E0000",  /* Hex digit 2 */
      "0033:000000003C4242021C020242423C0000",  /* Hex digit 3 */
      "0034:00000000040C142444447E0404040000",  /* Hex digit 4 */
      "0035:000000007E4040407C020202423C0000",  /* Hex digit 5 */
      "0036:000000001C2040407C424242423C0000",  /* Hex digit 6 */
      "0037:000000007E0202040404080808080000",  /* Hex digit 7 */
      "0038:000000003C4242423C424242423C0000",  /* Hex digit 8 */
      "0039:000000003C4242423E02020204380000",  /* Hex digit 9 */
      "0041:0000000018242442427E424242420000",  /* Hex digit A */
      "0042:000000007C4242427C424242427C0000",  /* Hex digit B */
      "0043:000000003C42424040404042423C0000",  /* Hex digit C */
      "0044:00000000784442424242424244780000",  /* Hex digit D */
      "0045:000000007E4040407C404040407E0000",  /* Hex digit E */
      "0046:000000007E4040407C40404040400000",  /* Hex digit F */
      "0055:000000004242424242424242423C0000",  /* Unicode 'U' */
      "002B:0000000000000808087F080808000000"   /* Unicode '+' */
   };
unsigned char hexbits[18][32]; /* The above digits converted into bitmap */

unsigned unipage=0;        /* Unicode page number, 0x00..0xff */
int flip=1;                /* transpose entire matrix as in Unicode book */



int
main (int argc, char *argv[])
{

   int i, j;                  /* loop variables                    */
   unsigned k0;               /* temp Unicode char variable        */
   unsigned swap;             /* temp variable for swapping values */
   char inbuf[256];           /* input buffer                      */
   unsigned filesize;         /* size of file in bytes             */
   unsigned bitmapsize;       /* size of bitmap image in bytes     */
   unsigned thischar;         /* the current character             */
   unsigned char thischarbyte; /* unsigned char lowest byte of Unicode char */
   int thischarrow;           /* row 0..15 where this character belongs  */
   int thiscol;               /* column 0..15 where this character belongs */
   int toppixelrow;           /* pixel row, 0..16*32-1               */
   unsigned lastpage=0;       /* the last Unicode page read in font file */
   int wbmp=0;                /* set to 1 if writing .wbmp format file */

   unsigned char bitmap[17*32][18*4]; /* final bitmap */
   unsigned char charbits[32][4];  /* bitmap for one character, 4 bytes/row */

   char *infile="", *outfile="";  /* names of input and output files */
   FILE *infp, *outfp;      /* file pointers of input and output files */

   int init();                  /* initializes bitmap row/col labeling, &c. */
   int hex2bit();               /* convert hex string --> bitmap */

   bitmapsize = 17*32*18*4;  /* 17 rows by 18 cols, each 4 bytes */

   if (argc > 1) {
      for (i = 1; i < argc; i++) {
         if (argv[i][0] == '-') {  /* this is an option argument */
            switch (argv[i][1]) {
               case 'f':  /* flip (transpose) glyphs in bitmap as in standard */
                  flip = !flip;
                  break;
               case 'i':  /* name of input file */
                  infile = &argv[i][2];
                  break;
               case 'o':  /* name of output file */
                  outfile = &argv[i][2];
                  break;
               case 'p':  /* specify a Unicode page other than default of 0 */
                  sscanf (&argv[i][2], "%x", &unipage); /* Get Unicode page */
                  break;
               case 'w':  /* write a .wbmp file instead of a .bmp file */
                  wbmp = 1;
                  break;
               default:   /* if unrecognized option, print list and exit */
                  fprintf (stderr, "\nSyntax:\n\n");
                  fprintf (stderr, "   %s -p<Unicode_Page> ", argv[0]);
                  fprintf (stderr, "-i<Input_File> -o<Output_File> -w\n\n");
                  fprintf (stderr, "   -w specifies .wbmp output instead of ");
                  fprintf (stderr, "default Windows .bmp output.\n\n");
                  fprintf (stderr, "   -p is followed by 1 to 6 ");
                  fprintf (stderr, "Unicode page hex digits ");
                  fprintf (stderr, "(default is Page 0).\n\n");
                  fprintf (stderr, "\nExample:\n\n");
                  fprintf (stderr, "   %s -p83 -iunifont.hex -ou83.bmp\n\n\n",
                         argv[0]);
                  exit (1);
            }
         }
      }
   }
   /*
      Make sure we can open any I/O files that were specified before
      doing anything else.
   */
   if (strlen (infile) > 0) {
      if ((infp = fopen (infile, "r")) == NULL) {
         fprintf (stderr, "Error: can't open %s for input.\n", infile);
         exit (1);
      }
   }
   else {
      infp = stdin;
   }
   if (strlen (outfile) > 0) {
      if ((outfp = fopen (outfile, "w")) == NULL) {
         fprintf (stderr, "Error: can't open %s for output.\n", outfile);
         exit (1);
      }
   }
   else {
      outfp = stdout;
   }

   (void)init(bitmap); /* initialize bitmap with row/column headers, etc. */

   /*
      Read in the characters in the page
   */
   while (lastpage <= unipage && fgets (inbuf, MAXBUF-1, infp) != NULL) {
      sscanf (inbuf, "%x", &thischar);
      lastpage = thischar >> 8; /* keep Unicode page to see if we can stop */
      if (lastpage == unipage) {
         thischarbyte = (unsigned char)(thischar & 0xff);
         for (k0=0; inbuf[k0] != ':'; k0++);
         k0++;
         hex2bit (&inbuf[k0], charbits);  /* convert hex string to 32*4 bitmap */

         /*
            Now write character bitmap upside-down in page array, to match
            .bmp file order.  In the .wbmp` and .bmp files, white is a '1'
            bit and black is a '0' bit, so complement charbits[][].
         */

         thiscol = (thischarbyte & 0xf) + 2;  /* column number will be 1..16  */
         thischarrow = thischarbyte >> 4;     /* charcter row number, 0..15   */
         if (flip) {  /* swap row and column placement */
            swap = thiscol;
            thiscol = thischarrow;
            thischarrow = swap;
            thiscol += 2;       /* column index starts at 1 */
            thischarrow -= 2;   /* row index starts at 0    */
         }
         toppixelrow = 32 * (thischarrow + 1) - 1; /* from bottom to top    */
   
         /*
            Copy the center of charbits[][] because hex characters only
            occupy rows 8 to 23 and column byte 2 (and for 16 bit wide
            characters, byte 3).  The charbits[][] array was given 32 rows
            and 4 column bytes for completeness in the beginning.
         */
         for (i=8; i<24; i++) {
            bitmap[toppixelrow + i][(thiscol << 2) | 0] =
               ~charbits[i][0] & 0xff;
            bitmap[toppixelrow + i][(thiscol << 2) | 1] =
               ~charbits[i][1] & 0xff;
            bitmap[toppixelrow + i][(thiscol << 2) | 2] =
               ~charbits[i][2] & 0xff;
            /* Only use first 31 bits; leave vertical rule in 32nd column */
            bitmap[toppixelrow + i][(thiscol << 2) | 3] =
               ~charbits[i][3] & 0xfe;
         }
         /*
            Leave white space in 32nd column of rows 8, 14, 15, and 23
            to leave 16 pixel height upper, middle, and lower guides.
         */
         bitmap[toppixelrow +  8][(thiscol << 2) | 3] |= 1;
         bitmap[toppixelrow + 14][(thiscol << 2) | 3] |= 1;
         bitmap[toppixelrow + 15][(thiscol << 2) | 3] |= 1;
         bitmap[toppixelrow + 23][(thiscol << 2) | 3] |= 1;
      }
   }
   /*
      Now write the appropriate bitmap file format, either
      Wireless Bitmap or Microsoft Windows bitmap.
   */
   if (wbmp) {  /* Write a Wireless Bitmap .wbmp format file */
      /*
         Write WBMP header
      */
      fprintf (outfp, "%c", 0x00); /* Type of image; always 0 (monochrome) */
      fprintf (outfp, "%c", 0x00); /* Reserved; always 0                   */
      fprintf (outfp, "%c%c", 0x84, 0x40); /* Width  = 576 pixels          */
      fprintf (outfp, "%c%c", 0x84, 0x20); /* Height = 544 pixels          */
      /*
         Write bitmap image
      */
      for (toppixelrow=0; toppixelrow <= 17*32-1; toppixelrow++) {
         for (j=0; j<18; j++) {
            fprintf (outfp, "%c", bitmap[toppixelrow][(j<<2)    ]);
            fprintf (outfp, "%c", bitmap[toppixelrow][(j<<2) | 1]);
            fprintf (outfp, "%c", bitmap[toppixelrow][(j<<2) | 2]);
            fprintf (outfp, "%c", bitmap[toppixelrow][(j<<2) | 3]);
         }
      }
   }
   else {  /* otherwise, write a Microsoft Windows .bmp format file */
      /*
         Write the .bmp file -- start with the header, then write the bitmap
      */

      /* 'B', 'M' appears at start of every .bmp file */
      fprintf (outfp, "%c%c", 0x42, 0x4d);

      /* Write file size in bytes */
      filesize   = 0x3E + bitmapsize;
      fprintf (outfp, "%c", (unsigned char)((filesize        ) & 0xff));
      fprintf (outfp, "%c", (unsigned char)((filesize >> 0x08) & 0xff));
      fprintf (outfp, "%c", (unsigned char)((filesize >> 0x10) & 0xff));
      fprintf (outfp, "%c", (unsigned char)((filesize >> 0x18) & 0xff));

      /* Reserved - 0's */
      fprintf (outfp, "%c%c%c%c", 0x00, 0x00, 0x00, 0x00);

      /* Offset from start of file to bitmap data */
      fprintf (outfp, "%c%c%c%c", 0x3E, 0x00, 0x00, 0x00);

      /* Length of bitmap info header */
      fprintf (outfp, "%c%c%c%c", 0x28, 0x00, 0x00, 0x00);

      /* Width of bitmap in pixels */
      fprintf (outfp, "%c%c%c%c", 0x40, 0x02, 0x00, 0x00);

      /* Height of bitmap in pixels */
      fprintf (outfp, "%c%c%c%c", 0x20, 0x02, 0x00, 0x00);

      /* Planes in bitmap (fixed at 1) */
      fprintf (outfp, "%c%c", 0x01, 0x00);

      /* bits per pixel (1 = monochrome) */
      fprintf (outfp, "%c%c", 0x01, 0x00);

      /* Compression (0 = none) */
      fprintf (outfp, "%c%c%c%c", 0x00, 0x00, 0x00, 0x00);

      /* Size of bitmap data in bytes */
      fprintf (outfp, "%c", (unsigned char)((bitmapsize        ) & 0xff));
      fprintf (outfp, "%c", (unsigned char)((bitmapsize >> 0x08) & 0xff));
      fprintf (outfp, "%c", (unsigned char)((bitmapsize >> 0x10) & 0xff));
      fprintf (outfp, "%c", (unsigned char)((bitmapsize >> 0x18) & 0xff));

      /* Horizontal resolution in pixels per meter */
      fprintf (outfp, "%c%c%c%c", 0xC4, 0x0E, 0x00, 0x00);

      /* Vertical resolution in pixels per meter */
      fprintf (outfp, "%c%c%c%c", 0xC4, 0x0E, 0x00, 0x00);

      /* Number of colors used */
      fprintf (outfp, "%c%c%c%c", 0x02, 0x00, 0x00, 0x00);

      /* Number of important colors */
      fprintf (outfp, "%c%c%c%c", 0x02, 0x00, 0x00, 0x00);

      /* The color black: B=0x00, G=0x00, R=0x00, Filler=0xFF */
      fprintf (outfp, "%c%c%c%c", 0x00, 0x00, 0x00, 0x00);

      /* The color white: B=0xFF, G=0xFF, R=0xFF, Filler=0xFF */
      fprintf (outfp, "%c%c%c%c", 0xFF, 0xFF, 0xFF, 0x00);

      /*
         Now write the raw data bits.  Data is written from the lower
         left-hand corner of the image to the upper right-hand corner
         of the image.
      */
      for (toppixelrow=17*32-1; toppixelrow >= 0; toppixelrow--) {
         for (j=0; j<18; j++) {
            fprintf (outfp, "%c", bitmap[toppixelrow][(j<<2)    ]);
            fprintf (outfp, "%c", bitmap[toppixelrow][(j<<2) | 1]);
            fprintf (outfp, "%c", bitmap[toppixelrow][(j<<2) | 2]);

            fprintf (outfp, "%c", bitmap[toppixelrow][(j<<2) | 3]);
         }
      }
   }
   exit (0);
}

/*
   Convert the portion of a hex string after the ':' into a character bitmap.

   If string is >= 128 characters, it will fill all 4 bytes per row.
   If string is >= 64 characters and < 128, it will fill 2 bytes per row.
   Otherwise, it will fill 1 byte per row.
*/
int
hex2bit (char *instring, unsigned char character[32][4])
{

   int i;  /* current row in bitmap character */
   int j;  /* current character in input string */
   int k;  /* current byte in bitmap character */
   int width;  /* number of output bytes to fill - 1: 0, 1, 2, or 3 */

   for (i=0; i<32; i++)  /* erase previous character */
      character[i][0] = character[i][1] = character[i][2] = character[i][3] = 0; 
   j=0;  /* current location is at beginning of instring */

   if (strlen (instring) <= 34)  /* 32 + possible '\r', '\n' */
      width = 0;
   else if (strlen (instring) <= 66)  /* 64 + possible '\r', '\n' */
      width = 1;
   else if (strlen (instring) <= 98)  /* 96 + possible '\r', '\n' */
      width = 3;
   else  /* the maximum allowed is quadruple-width */
      width = 4;

   k = (width > 1) ? 0 : 1; /* if width > double, start at index 1 else at 0 */

   for (i=8; i<24; i++) {  /* 16 rows per input character, rows 8..23 */
      sscanf (&instring[j], "%2hhx", &character[i][k]);
      j += 2;
      if (width > 0) { /* add next pair of hex digits to this row */
         sscanf (&instring[j], "%2hhx", &character[i][k+1]);
         j += 2;
         if (width > 1) { /* add next pair of hex digits to this row */
            sscanf (&instring[j], "%2hhx", &character[i][k+2]);
            j += 2;
            if (width > 2) {  /* quadruple-width is maximum width */
               sscanf (&instring[j], "%2hhx", &character[i][k+3]);
               j += 2;
            }
         }
      }
   }

   return (0);
}

int
init (unsigned char bitmap[17*32][18*4])
{
   int i, j;
   unsigned char charbits[32][4];  /* bitmap for one character, 4 bytes/row */
   unsigned toppixelrow;
   unsigned thiscol;
   unsigned char pnybble0, pnybble1, pnybble2, pnybble3;

   for (i=0; i<18; i++) { /* bitmaps for '0'..'9', 'A'-'F', 'u', '+' */

      hex2bit (&hex[i][5], charbits);  /* convert hex string to 32*4 bitmap */

      for (j=0; j<32; j++) hexbits[i][j] = ~charbits[j][1];
   }

   /*
      Initialize bitmap to all white.
   */
   for (toppixelrow=0; toppixelrow < 17*32; toppixelrow++) {
      for (thiscol=0; thiscol<18; thiscol++) {
         bitmap[toppixelrow][(thiscol << 2)    ] = 0xff;
         bitmap[toppixelrow][(thiscol << 2) | 1] = 0xff;
         bitmap[toppixelrow][(thiscol << 2) | 2] = 0xff;
         bitmap[toppixelrow][(thiscol << 2) | 3] = 0xff;
      }
   }
   /*
      Write the "u+nnnn" table header in the upper left-hand corner,
      where nnnn is the upper 16 bits of a 32-bit Unicode assignment.
   */
   pnybble3 = (unipage >> 20);
   pnybble2 = (unipage >> 16) & 0xf;
   pnybble1 = (unipage >> 12) & 0xf;
   pnybble0 = (unipage >>  8) & 0xf;
   for (i=0; i<32; i++) {
      bitmap[i][1] = hexbits[16][i];  /* copy 'u' */
      bitmap[i][2] = hexbits[17][i];  /* copy '+' */
      bitmap[i][3] = hexbits[pnybble3][i];
      bitmap[i][4] = hexbits[pnybble2][i];
      bitmap[i][5] = hexbits[pnybble1][i];
      bitmap[i][6] = hexbits[pnybble0][i];
   }
   /*
      Write low-order 2 bytes of Unicode number assignments, as hex labels
   */
   pnybble3 = (unipage >> 4) & 0xf;  /* Highest-order hex digit */
   pnybble2 = (unipage     ) & 0xf;  /* Next highest-order hex digit */
   /*
      Write the column headers in bitmap[][] (row headers if flipped)
   */
   toppixelrow = 32 * 17 - 1; /* maximum pixel row number */
   /*
      Label the column headers. The hexbits[][] bytes are split across two
      bitmap[][] entries to center a the hex digits in a column of 4 bytes.
      OR highest byte with 0xf0 and lowest byte with 0x0f to make outer
      nybbles white (0=black, 1-white).
   */
   for (i=0; i<16; i++) {
      for (j=0; j<32; j++) {
         if (flip) {  /* transpose matrix */
            bitmap[j][((i+2) << 2) | 0]  = (hexbits[pnybble3][j] >> 4) | 0xf0;
            bitmap[j][((i+2) << 2) | 1]  = (hexbits[pnybble3][j] << 4) |
                                           (hexbits[pnybble2][j] >> 4);
            bitmap[j][((i+2) << 2) | 2]  = (hexbits[pnybble2][j] << 4) |
                                           (hexbits[i][j] >> 4);
            bitmap[j][((i+2) << 2) | 3]  = (hexbits[i][j] << 4) | 0x0f;
         }
         else {
            bitmap[j][((i+2) << 2) | 1] = (hexbits[i][j] >> 4) | 0xf0;
            bitmap[j][((i+2) << 2) | 2] = (hexbits[i][j] << 4) | 0x0f;
         }
      }
   }
   /*
      Now use the single hex digit column graphics to label the row headers.
   */
   for (i=0; i<16; i++) {
      toppixelrow = 32 * (i + 1) - 1; /* from bottom to top    */

      for (j=0; j<32; j++) {
         if (!flip) {  /* if not transposing matrix */
            bitmap[toppixelrow + j][4] = hexbits[pnybble3][j];
            bitmap[toppixelrow + j][5] = hexbits[pnybble2][j];
         }
         bitmap[toppixelrow + j][6] = hexbits[i][j];
      }
   }
   /*
      Now draw grid lines in bitmap, around characters we just copied.
   */
   /* draw vertical lines 2 pixels wide */
   for (i=1*32; i<17*32; i++) {
      if ((i & 0x1f) == 7)
         i++;
      else if ((i & 0x1f) == 14)
         i += 2;
      else if ((i & 0x1f) == 22)
         i++;
      for (j=1; j<18; j++) {
         bitmap[i][(j << 2) | 3] &= 0xfe;
      }
   }
   /* draw horizontal lines 1 pixel tall */
   for (i=1*32-1; i<18*32-1; i+=32) {
      for (j=2; j<18; j++) {
         bitmap[i][(j << 2)    ] = 0x00;
         bitmap[i][(j << 2) | 1] = 0x81;
         bitmap[i][(j << 2) | 2] = 0x81;
         bitmap[i][(j << 2) | 3] = 0x00;
      }
   }
   /* fill in top left corner pixel of grid */
   bitmap[31][7] = 0xfe;

   return (0);
}
